File: crystal_lattice.py
def check(row):
    for i in range(1, len(row)):
        if row[i] == row[i-1]:
            return True
    return False

def golf(cube):
    x = [check(row) for grid in cube for row in grid]
    x += [check(row) for grid in cube for row in zip(*grid)]
    x += [check(row) for grid in zip(*cube) for row in zip(*grid)]
    return not any(x)

if __name__ == "__main__":
    assert golf([[["a", "b"],
                  ["c", "d"]],
                 [["e", "f"],
                  ["g", "h"]]]) == True, "st example"
 
    assert golf([[["X", "Z"],
                  ["Z", "X"]],
                 [["Z", "X"],
                  ["X", "Z"]]]) == True, "1st example"
    assert golf([[["X", "Z"],
                  ["Z", "X"]],
                 [["X", "Z"],
                  ["Z", "X"]]]) == False, "2nd example"
    print("All done? Earn rewards by using the 'Check' button!")

File: crystal_lattice.js
function f(x){
  var z=0;
  for(var i in x){
    if(z === x[i]) return 1
    z = x[i]
  }
  return 0
}
function golf(g){
  var m = g.map((_,c)=>g.map(r=>r[c]))
  for (var i in g) if(f(g[i])) return 0
  for (var i in m) if(f(m[i])) return 0
  for (var i in m) {
    var x = m[i].map((_,c)=>m[i].map(r=>r[c]))
    for (var j in x) if(f(x[j])) return 0
  }
  return 1
}

var assert = require('assert');

if (!global.is_checking) {
  // These "asserts" using only for self-checking and not necessary for auto-testing
  assert.equal(golf([[["X", "Z"],
    ["Z", "X"]],
    [["Z", "X"],
      ["X", "Z"]]]), true, "1st example)");
  assert.equal(golf([[["X", "Z"],
    ["Z", "X"]],
    [["X", "Z"],
      ["Z", "X"]]]), false, "2nd example)");
  console.log("Now that you're finished, hit the 'Check' button to review your code and earn sweet rewards!");
}
